src/rgw/rgw_op.cc

```diff
@@ -3514,3 +3514,79 @@ void RGWHandler::put_op(RGWOp *op)
   delete op;
 }
 
+
+
+/* Object Rename operation */
+
+int RGWRenameObj::verify_permission()
+{
+    // This function used to check ACLs in legacy code
+    // DSS does not require this.
+    return 0;
+}
+
+void RGWRenameObj::pre_exec()
+{
+  rgw_bucket_object_pre_exec(s);
+}
+
+void RGWRenameObj::execute()
+{
+    ret = 0;
+    rgw_obj_key orig_object;
+    orig_object.dss_duplicate(&(s->object));
+    (s->object).name = s->info.args.get("newname");
+    string copysource = s->bucket_name_str;
+    copysource.append("/");
+    copysource.append(orig_object.name);
+    ldout(s->cct, 0) << "DSS INFO: Converting to copy request. s->object: "
+                     << (s->object).name << ". Copy source: " << copysource << dendl;
+    s->info.env->set("HTTP_X_JCS_COPY_SOURCE", copysource.c_str());
+    s->info.env->set("HTTP_X_JCS_METADATA_DIRECTIVE", "COPY");
+    s->copy_source = s->info.env->get("HTTP_X_JCS_COPY_SOURCE");
+    RGWCopyObj_ObjStore_S3* copy_op = new RGWCopyObj_ObjStore_S3;
+    if (s->copy_source) {
+      ret = RGWCopyObj::parse_copy_location(s->copy_source, s->src_bucket_name, s->src_object);
+      if (!ret) {
+        ldout(s->cct, 0) << "DSS INFO: Rename op failed to parse copy location" << dendl;
+        return;
+      }
+      /*ret = validate_bucket_name(s->src_bucket_name, bucket_name_strictness_value);
+      if (ret) {
+        ldout(s->cct, 0) << "DSS INFO: Rename op bucket name not valid" << dendl;
+        return;
+      }
+      ret = validate_object_name(s->src_object.name);
+      if (ret) {
+        ldout(s->cct, 0) << "DSS INFO: Rename op object name not valid" << dendl;
+        return;
+      }*/
+    }
+
+    copy_op->init(store, s, dialect_handler);
+    copy_op->verify_op_mask();
+    copy_op->verify_permission();
+    copy_op->verify_params();
+    copy_op->init_processing();
+    s->system_request = true;
+    copy_op->pre_exec();
+    copy_op->execute();
+    ldout(s->cct, 0) << "DSS INFO: Rename op copy done" << dendl;
+
+    s->system_request = false;
+    (s->object).dss_duplicate(&orig_object);
```

Comment:
Shouldn't instead of name orig_object, you should use buffer_object, as you are first assigning the source object from s to orig_object, then assigning newname to s and then copying, and after copying you are assigning original key_name from orig_obj to s